README
This utility, gpsread, started as small program to pull NMEA data off a ND-100s
dongle, with the output going to a shell script. It got tidied up and made
more general, less hard-compiled defaults. As such it was an excercise in
learning libConfuse configuration parsing library.

I've released it under the GPLv2 and the BSD 2 clause license. Pick one.

Installation:
  $ make
  $ sudo make install
By default the installation prefix, set in Makefile, is /usr/local And so this
will install $(PREFIX)/bin/gpsread and $(PREFIX)/man/man1/gpsread.1.gz

A typical NMEA sentence that this utility expects:
  $GPGGA,170643.000,5237.7238,N,00115.1283,E,1,04,7.4,26.1,M,47.0,M,,0000*6A

This has been tested on 1 dongle, on 4 different operating systems, at 3 locations.
I wouldn't use it for your-life-may-depend-on-it navigation, and you shouldn't either.

W.B.Hill <mail@wbh.org> 20140921


ChangeLog
ChangeLog for gpsread.
=====================

0.9.3 gpsread-20170909
        Fixed a typo.
0.9.2   Bugfix for display of baudrate in usage() and help() messages.
        Added RadioHams' Maidenhead grid units.

0.9.1 gpsread-20150418
        Bugfix for baud rate setting.

0.9 gpsread-20140926
	Added OSGB output units.

0.8 gpsread-20140921
	Initial public release.
	W.B.Hill <mail@wbh.org>


osgb.c
#include <math.h>

// Written by Chuck Gantz - chuck.gantz@globalstar.com
// Compressed for AVR by Bill Hill - bugs@wbh.org

static const char GridSquare[] = "VWXYZQRSTULMNOPFGHJKABCDE" ;
static const double a = 6377563.396 ;
static const double k0 = 0.9996012717 ;
static const double M0 = 5429228.603180 ;
static const double eccS = 0.006670539761597 ;
static const double eccP = 0.006715334668516 ;
static const double lonOR = -0.034906585039887 ;
static const double latOR = +0.855211333477221 ;

// Converts lat/long to OSGB coords.
// Lat and Lon are in decimal degrees.
void
LLtoOSGB ( const double lat, const double lon, char* OSGBz, long* OSGBe, long* OSGBn )
  {
  long posx, posy ;
  double easting, northing ;
  double latR = lat*0.017453292519943 ;
  double lonR = lon*0.017453292519943 ;
  double N = a/sqrt(1-eccS*sin(latR)*sin(latR)) ;
  double T = tan(latR)*tan(latR) ;
  double C = eccP*cos(latR)*cos(latR) ;
  double A = cos(latR)*(lonR-lonOR) ;
  double M = a*(0.998330273507751*latR-0.002505636963581*sin(2*latR)+0.000002620236941*sin(4*latR)-0.000000003381659*sin(6*latR)) ;
  easting = 400000.0+(double)(k0*N*(A+(1-T+C)*A*A*A/6+(5-18*T+T*T+72*C-58*eccP)*A*A*A*A*A/120)) ;
  northing = (double)(k0*(M-M0+N*tan(latR)*(A*A/2+(5-T+9*C+4*C*C)*A*A*A*A/24+(61-58*T+T*T+600*C-330*eccP)*A*A*A*A*A*A/720)))-100000.0 ;
  *OSGBe = (long)(easting+0.5) ; *OSGBn = (long)(northing+0.5) ;
  posx = *OSGBe/500000L ; posy = *OSGBn/500000L ;
  OSGBz[0] = GridSquare[posx+posy*5+7] ;
  posx = *OSGBe%500000L ; posx = posx/100000L ;
  posy = *OSGBn%500000L ; posy = posy/100000L ;
  OSGBz[1] = GridSquare[posx+posy*5] ;
  OSGBz[2] = '\0' ;
  *OSGBn = *OSGBn%500000L ; *OSGBn = *OSGBn%100000L ;
  *OSGBe = *OSGBe%500000L ; *OSGBe = *OSGBe%100000L ;
  }


gpsread.h
#ifndef GPSREAD_H
#define GPSREAD_H


// Valid position units.
typedef enum { INVALID=0, TIME, NMEA, OSGB, MHEAD, LLMINSEC, LLMINDEC, LLDECIMAL } posunit_t ;
const char* posunit_names[] = { "INVALID", "TIME", "NMEA", "OSGB", "MHEAD", "LLMINSEC", "LLMINDEC", "LLDECIMAL" } ;

// Set compile-time defaults.
#define TIMEOUT 15
#define GPSBAUD B4800
#define POSUNIT LLDECIMAL
#if __APPLE__
  #define GPSTERM "/dev/tty.usbserial"
#elif __linux
  #define GPSTERM "/dev/ttyUSB0"
#else
  #error "Unable to determine default for GPSTERM"
#endif

// Converts lat/long to OSGB coords. Lat and Lon are in decimal degrees.
void LLtoOSGB ( const double lat, const double lon, char* OSGBz, long* OSGBe, long* OSGBn ) ;


#endif //GPSREAD_H


// dbext:profile=testdb
// vim:ts=2:sw=2:tw=150:fo=tcnq2b:foldmethod=indent


gpsread.c
/****************************************************************************************************************************************************/
/*  Purpose:    Read, parse and display data from a USB serial GPS.                                                                                 */
/*  Author:     Copyright (c) 2014, W.B.Hill <mail@wbh.org> All rights reserved.                                                                    */
/*  License:    GPLv2 - see file LICENSE or http://www.gnu.org                                                                                      */
/*  License:    BSD - see http://opensource.org/licenses/BSD-2-Clause                                                                               */
/****************************************************************************************************************************************************/
/*                                                                                                                                                  */
/*  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:  */
/*  --- Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.                */
/*  --- Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the       */
/*      documentation and/or other materials provided with the distribution.                                                                        */
/*                                                                                                                                                  */
/*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED   */
/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR    */
/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,       */
/*  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF       */
/*  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS         */
/*  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.                                                                                    */
/*                                                                                                                                                  */
/****************************************************************************************************************************************************/

// gcc -o gpsread gpsread.c -lintl -lconfuse
// New arguments need to be added where tagged 'ADDARG'.

// Standard infrastructure
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
// Trig functions.
#include <math.h>
// For serial comms.
#include <fcntl.h>
#include <termios.h>
// For basename()
#include <libgen.h>
// Options parsing.
#include <getopt.h>
// Config file reading.
#include <confuse.h>
// For timeouts.
#include <signal.h>
// Required by strtol()
#include <limits.h>
// Compile-time defults.
#include "gpsread.h"


// Convert macros to strings.
#define STR(S) _STR(S)
#define _STR(S) #S


// Die if we hit the timeout.
void
sighandler ( int sig )
  {
  if ( sig == SIGALRM )
    {
    fprintf ( stderr, "Timed out trying to read GPS.\n" ) ;
    exit ( EXIT_FAILURE ) ;
    }
  }


// Map a baudrate to a number.
int
map_baud ( int br )
  {
  switch ( br )
    {
    case B50 :
      return 50 ;
    case B75 :
      return 75 ;
    case B110 :
      return 110 ;
    case B134 :
      return 134 ;
    case B150 :
      return 150 ;
    case B200 :
      return 200 ;
    case B300 :
      return 300 ;
    case B600 :
      return 600 ;
    case B1200 :
      return 1200 ;
    case B1800 :
      return 1800 ;
    case B2400 :
      return 2400 ;
    case B4800 :
      return 4800 ;
    case B9600 :
      return 9600 ;
    case B19200 :
      return 19200 ;
    case B38400 :
      return 38400 ;
    default :
      return -1 ;
    }
  }


// Convert and check the baud rate is OK.
int
valid_baud ( int br )
  {
  switch ( br )
    {
    case 50 :
      return B50 ;
    case 75 :
      return B75 ;
    case 110 :
      return B110 ;
    case 134 :
      return B134 ;
    case 150 :
      return B150 ;
    case 200 :
      return B200 ;
    case 300 :
      return B300 ;
    case 600 :
      return B600 ;
    case 1200 :
      return B1200 ;
    case 1800 :
      return B1800 ;
    case 2400 :
      return B2400 ;
    case 4800 :
      return B4800 ;
    case 9600 :
      return B9600 ;
    case 19200 :
      return B19200 ;
    case 38400 :
      return B38400 ;
    default :
      return -1 ;
    }
  }


// Convert a string to a posuint_t
posunit_t
map_posunit ( const char* value )
  {
  if ( !strcasecmp ( value, posunit_names[TIME] ) ) return TIME ;
  else if ( !strcasecmp ( value, posunit_names[NMEA] ) ) return NMEA ;
  else if ( !strcasecmp ( value, posunit_names[OSGB] ) ) return OSGB ;
  else if ( !strcasecmp ( value, posunit_names[MHEAD] ) ) return MHEAD ;
  else if ( !strcasecmp ( value, posunit_names[LLMINSEC] ) ) return LLMINSEC ;
  else if ( !strcasecmp ( value, posunit_names[LLMINDEC] ) ) return LLMINDEC ;
  else if ( !strcasecmp ( value, posunit_names[LLDECIMAL] ) ) return LLDECIMAL ;
  else return INVALID ;
  }


// Check we don't get a negative.
int
validate_uint ( cfg_t* cfg, cfg_opt_t* opt )
  {
  int value = cfg_opt_getnint ( opt, 0 ) ;
  if ( value < 0 )
    {
    cfg_error ( cfg, "Invalid timeout." ) ;
    return -1 ;
    }
  return 0 ;
  }


// Check the baud rate's a good'un.
int
validate_baud ( cfg_t* cfg, cfg_opt_t* opt )
  {
  int value ;
  value = cfg_opt_getnint ( opt, 0 ) ;
  if ( valid_baud ( value ) == -1 )
    {
    cfg_error ( cfg, "Invalid baudrate." ) ;
    return -1 ;
    }
  return 0 ;
  }


// Check the GPS device.
int
validate_term ( cfg_t* cfg, cfg_opt_t* opt )
  {
  /*
  if ( access ( cfg_opt_getnstr ( opt, 0 ), F_OK ) )
    {
    cfg_error(cfg, "Setting %s, problem with GPS tty value: %s", opt->name, cfg_opt_getnstr ( opt, 0 ) ) ;
    return -1 ;
    }
  */
  // Not NULL.
  if ( cfg_opt_getnstr ( opt, 0 ) == NULL )
    {
    cfg_error(cfg, "Setting %s, no GPS tty value.", opt->name ) ;
    return -1 ;
    }
  // Needs to be "/dev/tty*", at least 8 chars.
  if( strlen ( cfg_opt_getnstr ( opt, 0 ) ) < 8 )
    {
    cfg_error(cfg, "Setting %s, problem with GPS tty value: %s", opt->name, cfg_opt_getnstr ( opt, 0 ) ) ;
    return -1 ;
    }
  return 0 ;
  }


// Callback for posunit in the config files.
int
parse_posunit ( cfg_t* cfg, cfg_opt_t* opt, const char* value, void* result )
  {
  // We don't use this.
  opt = opt ;
  // Allocate for the returned value.
  *(void**) result = malloc ( sizeof ( int ) ) ;
  // Save the answer.
  int mapped = map_posunit ( value ) ;
  // OK?
  if ( mapped )
    {
    **(posunit_t**) result = mapped ;
    return 0 ;
    }
  else
    {
    cfg_error ( cfg, "Invalid position unit '%s'", value ) ;
    free ( *(void**) result ) ;
    return 1 ;
    }
  }


// Show terse info. ADDARG
void
usage ( char* appname )
  {
  printf ( "Usage: %s -t%d -b%d -d%s -u%s\n", appname, TIMEOUT, map_baud(GPSBAUD), GPSTERM, STR(POSUNIT) ) ;
  }


// Show version info.
void
version ( char* appname )
  {
  printf ( "%s v%s, W.B.Hill <mail@wbh.org>, 19 Sept 2014\n", appname, STR(VERSION) ) ;
  }


// Show some help. ADDARG
void
help ( char* appname )
  {
  printf ( "Usage: %s [option] ...\n", appname ) ;
  printf ( "\t-h,--help     This help.\n" ) ;
  printf ( "\t-t,--timeout  Time to wait for GPS in seconds, default %ds\n", TIMEOUT ) ;
  printf ( "\t-b,--baudrate GPS device baudrate. Default %d\n", map_baud(GPSBAUD) ) ;
  printf ( "\t-d,--device   GPS tty device. Default %s\n", GPSTERM ) ;
  printf ( "\t-u,--units    Units to show position in. Default %s\n", STR(POSUNIT) ) ;
  printf ( "%s v%s, W.B.Hill <mail@wbh.org>, 19 Sept 2014\n", appname, STR(VERSION) ) ;
  }


// Parse the config files, then the command line, setup and then finally do stuff.
int
main ( int argc, char* argv[] )
  {
  // Configuration variables. ADDARG
  static int timeout ;
  static int gpsbaud ;
  static char* gpsterm ;
  static posunit_t posunit ;
  // Config file. ADDARG
  static cfg_opt_t opts[] =
    {
    CFG_INT ( "timeout", TIMEOUT, CFGF_NONE ),
    CFG_INT ( "gpsbaud", GPSBAUD, CFGF_NONE ),
    CFG_STR ( "gpsterm", GPSTERM, CFGF_NONE ),
    CFG_PTR_CB ( "posunit", STR(POSUNIT), CFGF_NONE, parse_posunit, free ),
    CFG_END()
    } ;
  // Command line options. ADDARG
  static struct option long_options[] =
    {
      { "help",      no_argument,       0,  'h' },
      { "version",   no_argument,       0,  'v' },
      { "timeout",   required_argument, 0,  't' },
      { "baudrate",  required_argument, 0,  'b' },
      { "device",    required_argument, 0,  'd' },
      { "units",     required_argument, 0,  'u' },
      { 0, 0, 0, 0 }
    } ;
  // Load the config files.
  char* etcconf ;
  char* usrconf ;
  cfg_t* confuse ;
  // Generate the config file names from the app name.
  asprintf ( &etcconf, "/etc/%s.conf", basename ( argv[0] ) ) ;
  asprintf ( &usrconf, "%s/.%src", getenv ( "HOME" ), basename ( argv[0] ) ) ;
  // Init the config library.
  confuse = cfg_init ( opts, CFGF_NONE ) ;
  // Check the values.
  cfg_set_validate_func ( confuse, "timeout", validate_uint ) ;
  cfg_set_validate_func ( confuse, "gpsbaud", validate_baud ) ;
  cfg_set_validate_func ( confuse, "gpsterm", validate_term ) ;
  // Read the /etc/app.conf file.
  if ( cfg_parse ( confuse, etcconf ) == CFG_PARSE_ERROR )
    {
    fprintf ( stderr, "Problem with config file '%s'\n", etcconf ) ;
    exit ( EXIT_FAILURE ) ;
    }
  // Read the ~/.apprc file.
  if ( cfg_parse ( confuse, usrconf ) == CFG_PARSE_ERROR )
    {
    fprintf ( stderr, "Problem with config file '%s'\n", usrconf ) ;
    exit ( EXIT_FAILURE ) ;
    }
  // Save the values. ADDARG
  gpsterm = strdup ( cfg_getstr ( confuse, "gpsterm" ) ) ;
  gpsbaud = cfg_getint ( confuse, "gpsbaud" ) ;
  timeout = cfg_getint ( confuse, "timeout" ) ;
  posunit = *(posunit_t*) cfg_getptr ( confuse, "posunit" ) ;
  // Done - free stuff.
  cfg_free ( confuse ) ;
  free ( etcconf ) ;
  free ( usrconf ) ;
  // Now parse the command line.
  int opt = 0 ;
  int long_index = 0 ;
  // Process the command line ADDARG
  while ( ( opt = getopt_long ( argc, argv, "hvt:b:d:u:", long_options, &long_index ) ) != -1 )
    {
    switch ( opt )
      {
      case 'h' :
        help ( basename ( argv[0] ) ) ;
        exit ( EXIT_SUCCESS ) ;
      case 'v' :
        version ( basename ( argv[0] ) ) ;
        exit ( EXIT_SUCCESS ) ;
      case 't' :
        timeout = (int) strtol ( optarg, (char **)NULL, 10 ) ;
        break ;
      case 'b' :
        gpsbaud = valid_baud ( (int) strtol ( optarg, (char **)NULL, 10 ) ) ;
        if ( gpsbaud == -1  )
          {
          fprintf ( stderr, "Invalid baudrate: %s\n", optarg ) ;
          exit ( EXIT_FAILURE ) ;
          }
        break ;
      case 'd' :
        // Needs to be "/dev/tty*", at least 8 chars.
        if( strlen ( optarg ) < 8 )
          {
          fprintf ( stderr, "Problem with GPS tty value: %s\n", optarg ) ;
          exit ( EXIT_FAILURE ) ;
          }
        else
          {
          free ( gpsterm ) ;
          gpsterm = strdup ( optarg  ) ;
          }
        break ;
      case 'u' :
        if ( map_posunit ( optarg ) )
          {
          posunit =  map_posunit ( optarg ) ;
          }
        else
          {
          fprintf ( stderr, "Invalid position unit: %s\n", optarg ) ;
          exit ( EXIT_FAILURE ) ;
          }
        break ;
      default :
        usage ( basename ( argv[0] ) ) ;
        exit ( EXIT_FAILURE ) ;
      }
    }
  // Shouldn't be anything left.
  if ( optind != argc )
    {
    fprintf ( stderr, "Unexpected command line argument.\n" ) ;
    usage ( basename ( argv[0] ) ) ;
    exit ( EXIT_FAILURE ) ;
    }
  // Set a callback for the alarm() signal.
  signal ( SIGALRM, sighandler ) ;
  // Timeout after specified seconds.
  alarm ( timeout ) ;
  // Start serial comms to GPS.
  int tty ;
  struct termios gpsio ;
  // Default to zero.
  memset ( &gpsio, 0, sizeof ( gpsio ) ) ;
  // Set for GPS comms.
  gpsio.c_iflag = 0 ;
  gpsio.c_oflag = 0 ;
  gpsio.c_cflag = CS8 | CREAD | CLOCAL ;
  gpsio.c_lflag = 0 ;
  gpsio.c_cc[VMIN] = 1 ;
  gpsio.c_cc[VTIME] = 5 ;
  // Open the tty device.
  tty = open ( gpsterm, O_RDWR | O_NONBLOCK ) ;
  // Did it work?
  if ( tty < 0 )
    {
    perror ( "Can't access GPS device" ) ;
    exit ( EXIT_FAILURE );
    }
  // Set baud rate.
  cfsetospeed ( &gpsio, gpsbaud ) ;
  cfsetispeed ( &gpsio, gpsbaud ) ;
  // Set properties.
  tcsetattr ( tty, TCSANOW, &gpsio ) ;
  // Attempt to fetch data.
  int i, len = 0, state = 0 ;
  unsigned char c ;
  char gpsbuffer[256] ;
  // The data.
  char utctime[10] ;
  char* raw ;
  char latc[3], lonc[4] ;
  int latd, lond ;
  double latm, lonm ;
  // Loop until break or timeout.
  while ( 1 )
    {
    // Read the GPSdongle
    i = read ( tty, &c, 1 ) ;
    // Give up if it's got nothing for us.
    if ( i <= 0 )
      {
      usleep ( 250 ) ;
      continue ;
      }
    // Don't save until the start of a sentence.
    if ( state == 0 )
      {
      if ( c == '$' ) state = 1 ;
      }
    // Save the data data.
    else if ( state == 1 )
      {
      // The end?
      if ( c == '\r' )
        {
        // Make it a valid string.
        gpsbuffer[len] = '\0' ;
        // Ready for the next time.
        len = 0 ;state = 0 ;
        // Process it.
        if ( ( gpsbuffer[0] == 'G' ) && ( gpsbuffer[1] == 'P' ) && ( gpsbuffer[2] == 'G' ) && ( gpsbuffer[3] == 'G' ) && ( gpsbuffer[4] == 'A' ) )
          {
          char **fp, *field[15], *ds ;
          ds = gpsbuffer + 6 ;
          // Save a copy, before it gets split.
          raw = strdup ( gpsbuffer ) ;
          for ( fp = field ; ( *fp = strsep( &ds, ",")) != NULL ; ) if ( ++fp >= &field[15]) break;
          /*
          0    = UTC of Position
          1    = Latitude
          2    = N or S
          3    = Longitude
          4    = E or W
          5    = GPS quality indicator (0=invalid; 1=GPS fix; 2=Diff. GPS fix)
          6    = Number of satellites in use [not those in view]
          7    = Horizontal dilution of position
          8    = Antenna altitude above/below mean sea level (geoid)
          9    = Meters  (Antenna height unit)
          10   = Geoidal separation (Diff. between WGS-84 earth ellipsoid and
                 mean sea level.  -=geoid is below WGS-84 ellipsoid)
          11   = Meters  (Units of geoidal separation)
          12   = Age in seconds since last update from diff. reference station
          13   = Diff. reference station ID#
          14   = Checksum
          */
          // Got a good reading?
          if ( (int) strtol ( field[5], (char **)NULL, 10 ) )
            {
            // Save it!
            snprintf ( utctime, 9, "%c%c:%c%c:%c%c", field[0][0], field[0][1], field[0][2], field[0][3], field[0][4], field[0][5] ) ;
            latc[0] = field[1][0] ;
            latc[1] = field[1][1] ;
            latc[2] = '\0' ;
            latd = (int) strtol ( latc, (char **)NULL, 10 ) ;
            latd *= ( field[2][0] == 'N' ) ? +1 : -1 ;
            latm = strtod ( field[1]+2, NULL ) ;
            lonc[0] = field[3][0] ;
            lonc[1] = field[3][1] ;
            lonc[2] = field[3][2] ;
            lonc[3] = '\0' ;
            lond = (int) strtol ( lonc, (char **)NULL, 10 ) ;
            lond *= ( field[4][0] == 'E' ) ? +1 : -1 ;
            lonm = strtod ( field[3]+3, NULL ) ;
            // Done it!
            break ;
            }
          else
            {
            free ( raw ) ;
            }
          }
        }
      // Save if not too much.
      else if ( len < 254 )
        {
        gpsbuffer[len++] = c ;
        }
      // This has gone on too long.
      else
        {
        // Give up.
        len = 0 ; state = 0 ;
        }
      }
    // Should never happen!
    else
      {
      // Give up.
      len = 0 ; state = 0 ;
      }
    }
  // Disable timeout.
  signal ( SIGALRM, SIG_IGN ) ;
  // Done with it now.
  close ( tty ) ;
  // Done with any config data. ADDARG
  free ( gpsterm ) ;
  // Show the data.
  double latt, lats, lont, lons ;
  char z[3] ;
  long e, n ;
  switch ( posunit )
    {
    case TIME :
      printf ( "%s\n", utctime ) ;
      break ;
    case NMEA :
      printf ( "$%s\n", raw ) ;
      break ;
    case OSGB :
      LLtoOSGB ( latd + latm / 60.0, lond + lonm / 60.0, z, &e, &n ) ;
      printf ( "[%s][%05ld][%05ld]\n", z, e, n ) ;
      break ;
    case MHEAD :
      latt =  90.0 + latd + latm / 60.0 ;
      lont = 180.0 + lond + lonm / 60.0 ;
      printf ( "%c", 'A' + (int) floor ( lont / 20.0 ) ) ;
      printf ( "%c", 'A' + (int) floor ( latt / 10.0 ) ) ;
      lont -= 20.0 * floor ( lont / 20.0 ) ;
      printf ( "%d", (int) floor ( lont / 2.0 ) ) ;
      latt -= 10.0 * floor ( latt / 10.0 ) ;
      printf ( "%d", (int) floor ( latt ) ) ;
      lont -= 2.0 * floor ( lont / 2.0 ) ;
      printf ( "%c", 'a' + (int) floor ( 12.0 * lont ) ) ;
      latt -= 1.0 * floor ( latt / 1.0 ) ;
      printf ( "%c", 'a' + (int) floor ( 24.0 * latt ) ) ;
      printf ( "\n" ) ;
      break ;
    case LLMINSEC :
      lats = 60.0 * modf ( latm, &latt ) ;
      printf ( "lat: %3d%c%02d\'%06.3f\"\n", latd, ( latd < 0 ) ? 'S' : 'N', ( int ) latt, lats ) ;
      lons = 60.0 * modf ( lonm, &lont ) ;
      printf ( "lon: %3d%c%02d\'%06.3f\"\n", lond, ( lond < 0 ) ? 'W' : 'E', ( int ) lont, lons ) ;
      break ;
    case LLMINDEC :
      printf ( "lat: %3d%c%8.4f\'\n", abs ( latd ), ( latd < 0 ) ? 'S' : 'N', latm ) ;
      printf ( "lon: %3d%c%8.4f\'\n", abs ( lond ), ( lond < 0 ) ? 'W' : 'E', lonm ) ;
      break ;
    case LLDECIMAL :
      printf ( "lat: %+10.5f\n", latd + latm / 60.0 ) ;
      printf ( "lon: %+10.5f\n", lond + lonm / 60.0 ) ;
      break ;
    default :
      // What happend here?
      break ;
    }
  free ( raw ) ;
  // That's all, folks!
  return EXIT_SUCCESS ;
  }


// VIM formatting info.
// vim:ts=2:sw=2:tw=150:fo=tcnq2b:foldmethod=indent
